QtAcademyProject.cpp

```diff
@@ -12,22 +12,155 @@
 
 #include "QtAcademyProject.h"
 
+#include "ArcGISVectorTiledLayer.h"
+#include "Basemap.h"
+#include "Error.h"
+#include "ExportTileCacheTask.h"
+#include "ExportVectorTilesJob.h"
+#include "ExportVectorTilesParameters.h"
+#include "ExportVectorTilesTask.h"
+#include "GeometryEngine.h"
+#include "Graphic.h"
+#include "GraphicListModel.h"
+#include "GraphicsOverlay.h"
+#include "GraphicsOverlayListModel.h"
+#include "ItemResourceCache.h"
+#include "LayerListModel.h"
+#include "Location.h"
+#include "LocationDisplay.h"
 #include "Map.h"
-#include "MapTypes.h"
 #include "MapQuickView.h"
+#include "MapTypes.h"
+#include "MapViewTypes.h"
+#include "OfflineMapTask.h"
+#include "Point.h"
+#include "Polygon.h"
+#include "PolylineBuilder.h"
+#include "SimpleLineSymbol.h"
+#include "SpatialReference.h"
+#include "SymbolTypes.h"
+#include "VectorTileCache.h"
+
+#include <QDateTime>
+#include <QDir>
+#include <QDirIterator>
+#include <QFile>
+#include <QFuture>
+#include <QNetworkInformation>
+#include <QStandardPaths>
+
 
 using namespace Esri::ArcGISRuntime;
 
+const QString vtpkPath = QStandardPaths::writableLocation(QStandardPaths::AppDataLocation);
+
 QtAcademyProject::QtAcademyProject(QObject* parent /* = nullptr */):
   QObject(parent),
-  m_map(new Map(BasemapStyle::OsmStandard, this))
+  m_map(new Map(this)),
+  m_basemap(new Basemap(BasemapStyle::OsmStandard, this))
 {
+  m_map->setBasemap(m_basemap);
+  m_tempObject.reset(new QObject(this));
 }
 
 QtAcademyProject::~QtAcademyProject()
 {
 }
 
+void QtAcademyProject::toggleOffline(bool offline)
+{
+  if (offline)
+  {
+    // Set the basemap to a null basemap
+    // Use a unique pointer so it will get cleaned up when there are no references to it
+    m_map->setBasemap(new Basemap(m_tempObject.get()));
+    // Load the offline base layers from memory
+    loadOfflineBasemaps();
+  }
+  else
+  {
+    // Set the map to the online basemap
+    m_map->setBasemap(m_basemap);
+
+    // Delete any offline objects by reseting the parent
+    m_tempObject.reset(new QObject(this));
+  }
+}
+
+void QtAcademyProject::loadOfflineBasemaps()
+{
+  // Check to see if any offline data exists
+  if (QFile::exists(vtpkPath+"/vectorTiles.vtpk"))
+  {
+    // Create a VectorTileCache object from the offline data
+    m_vectorTileCache = new VectorTileCache(vtpkPath+"/vectorTiles.vtpk", m_tempObject.get());
```

Comment:
subjective stylistic nitpick, but I generally find it easier to read when there are spaces between the variables and operators (applies throughout)

```suggestion
    m_vectorTileCache = new VectorTileCache(vtpkPath + "/vectorTiles.vtpk", m_tempObject.get());
```